### ISSUE TITLE
[Discussion] Proc and Thread Model

### ISSUE THREAD

**Issue by raedwulf** [sustrik/libdill#39]
I split this out from the original discussion.

```
P1         |P2                       proc (Unsupported on Windows)
-----+-----+-----+-----
T1   |T2   |T3   |T4                 thread (Windows & *nix)
--+--+--+--+--+--+--+--
C1|C2|C3|C4|C5|C6|C7|C8              go(routines) (libdill)
```

Currently in libdill, the API hints at and I think does support nested `proc`.  Going for the `spawn`-like approach would remove that hierarchy.

I think this could be reflected in the API:

```C
/* function that parses information from the parameters to determine whether the execution of the application is a child, returns -1 for failure, 0 for non-child, 1 for child */
int procinit();
/* spawn a new process of this program */
int procspawn(const char *argv[]);
```

`procspawn` sets an environment variable to communicate an named-IPC to communicate between the main process and the child processes (this approach allows minimal conflict with the command line parsing by the user and this is also supported on Win32)

The application pattern would be:
```C
int main(int argc, char **argv) {
    int rc;
    if(!(rc = procinit()) {
        /* main application */
        int h = procspawn(argv); /* spawn a process */
        /* sleep. do stuff */
        hclose(h);
    } else {
        /* child */
        if(dill_slow(rc == -1)) {
            if(errno == ENOPERM) fprintf(stderr, "cannot access process IPC\n");
            else fprintf(stderr, "unknown error\n");
            return 1;
        }
        /* do stuff in child process */
    }
}
```

On an unrelated note, I noticed that libdill has a function `msleep`.  Is the naming a throwback to libmill?

**Comment by raedwulf** [comment]
The semantics of `procspawn` is that processes share a flat hierarchy (as opposed to the structured concurrency of coroutines).  So if you issue `procspawn` within a child, the call will fail.

**Comment by sustrik** [comment]
msleep: it's "millisecond sleep" similar to posix usleep or nanosleep.

**Comment by sustrik** [comment]
Hm, aren't we moving in circles here?

The only use case here is to take advantage of all cores on a multicore machine. My understanding is that this proc/thread thing is good only for a single thing: Sharing listening TCP socket among processes/threads and thus load-balancing connections among them. And that seams like a rather weak rationale for dragging in the proc/thread complexity.

Maybe we should rather look at how to do TCP connection load balancing in a simpler (and more flexible!) way.

An example (not necessarily the right solution): Helper binary that opens the TCP listening socket on the port user wants to serve. Any process on that machine can connect to the helper process via AF_UNIX and get a dup of the listening TCP socket via SCM_RIGHTS.

**Comment by raedwulf** [comment]
That makes sense. Thinking about it these functions are effectively wrapper functions that the user could effectively implement himself.

I'll consider making a utility library as an alternative to multiplexing. 

For threads,  there's the use case where libdill can safely interact with other threads without breaking.  This would let a libdill instance per thread...  I've not tested but my branch should already support that simply by isolating the state. 

I'm still working on the details but I think there's also a way of cheaply creating thread pools per set of coroutines (like GOMAXPROCS but more flexible).  There's a fast path for single threaded "pools" which mimics existing behaviour.   The use case for multithreaded pools would be for processing a large number of independent logic that do not touch a shared state often - e.g. Objects within a game for instance.  I think this might be viable since context switching is down to 6ns and potentially the game logic is more costly than the thread safe context switching.

**Comment by raedwulf** [comment]
Thread pools would need internal libdill mechanisms,  however,  because the scheduler will need to know how to migrate and there should be a warning sign in the docs that although libdill synchronises the coroutines safely,  the user will need to take care of shared resources themselves.

**Comment by raedwulf** [comment]
My current thoughts on threaded coroutine scheduling is through work-stealing.  Migrating coroutines is cheap but not as cheap as leaving them on a particular thread's queue. So the idea would be to implement work-stealing, that is, a thread that is starved of work can steal half of another thread's work.

**Comment by sustrik** [comment]
Ah, I haven't thought of games.

I was thinking in terms of microservices and in that context threads are toxic. (Mediocre performance on one hand and no isolation on the other. Worst of all worlds and at a hefty price as well.)

In a game though the cost of crossing the process boundary may be much higher that cost of crossing a thread boundary, thus not really an option. (Is that true though? If there's 1:1 thread:core mapping there would be no TLB switches etc.)

Well, whatever the case, my main concern is how to isolate the core of the library from the threading mess. The ideal state would be to have a clean library that knows nothing of threads, yet provides extension points that allow to build different kinds of threading/procing support.

AFAIU your idea is to take what today is libdill's global state, turn it into an object  and spawn one for each thread. That's pretty neat. If the operation of creating such object would be exposed as an API, user can launch a thread (or fork a process), then use the API to get runtime env for that thread etc.

**Comment by raedwulf** [comment]
> I was thinking in terms of microservices and in that context threads are toxic. (Mediocre performance on one hand and no isolation on the other. Worst of all worlds and at a hefty price as well.)

Yes this is true.  I was thinking about an application that needs network ability as well as some computation ability.

> In a game though the cost of crossing the process boundary may be much higher that cost of crossing a thread boundary, thus not really an option. (Is that true though? If there's 1:1 thread:core mapping there would be no TLB switches etc.)

Partially in response to the previous comment as well.  I was thinking that coroutines could also be considered tasks (in fact I think some coroutine libraries call coroutines tasks as well).  One of the problems that is faced by modern games is that they can't properly use multicore systems because they normally constrain a particular subsystem per thread (e.g. graphics one thread (mainly because the traditional APIs were heavily state-based, although things have recently changed with Vulkan/DX12), physics another thread, game logic another thread) leading to some threads being starved of work before the end of a rendering cycle.

The aim is to make a task-based system based on coroutines by allowing coroutines migrate to threads that have no work to do.  In the context of a strategy game, for example Civilization or Crusader Kings, there are normally hundreds of different state machines simulating the AI of characters or of "diplomacy" etc.  In typical game design, these end up being separate game objects which would translate neatly into the concept of coroutines - especially at the speed libdill is able to switch between them.

I think I've solved the problem about synchronising multiple coroutine-queues efficiently using the work-stealing approach.  The tasks that remain are:

- design the API for thread/threadpools
- splitting coroutine queues (synchonisation is done, just need to split the coroutine ready queue)
- notifying a starved thread that work is available (linux eventfd/unix pipes)
- synchronise handle creation - I currently have independent handles per thread... I think make handles unique over all threads would actually be more sensible because then checks can be made to see which handle belongs to which thread (group).
- make chan multithread aware - `go` uses this as the main/only synchonisation primitive for multi-threaded applications. I suggest that we have the existing channel for single-threaded, and a multi-producer multi-consumer thread-safe channel for others.

> AFAIU your idea is to take what today is libdill's global state, turn it into an object and spawn one for each thread. That's pretty neat. If the operation of creating such object would be exposed as an API, user can launch a thread (or fork a process), then use the API to get runtime env for that thread etc.

I need to do some tests, but I think my branch has isolated the state per thread, I just need to provide some API to access it.  I was wondering, however, whether we should have a `dill_init` function, because at the moment, I have to check for initialisation in each exposed function in `libdill`.

**Comment by raedwulf** [comment]
Slightly related to threads, just clarifying the intended usage libdill.
Is the normal approach if you want to block a parent from terminating before its children either through `hclose` (for a quick termination) or safely through channels (to emulate a join-like behaviour)?

```C
int main() {
    int ch = channel(sizeof(int), 16);
    int h = go(worker(ch));
    chrecv(ch, NULL, 0, -1);
    hclose(h);
}
```

**Comment by sustrik** [comment]
Yes. The rule is:

1. Do hard-cancel by calling hclose() -- this is non blocking in fact as the coroutine cannot perform any blocking operations after is was hclosed and before it exits
2. Do soft-cancel by hand (typically some handshaking pattern using channels)

**Comment by raedwulf** [comment]
My current thoughts for an API (not fixed yet) are something along these lines:
```C
int ch = channel(sizeof(int), 16);
int h = go(worker(ch));
int t = thread(h);
chrecv(ch, NULL, 0, -1);
hclose(t);
```

I think those semantics for `hclose` should work the same - it just has to explicitly block until t returns.

In `go`, the scheduler is multithreaded using a threadpool with `GOMAXPROCS` different threads.

I'm aiming for a more flexible approach where a group of coroutines can run on the same thread pool.
```C
int s = scheduler(4); /* immediately spawn 4 threads but make them wait until work is available */
int h = go(worker(ch));
int rc = setsched(h, s); /* add the coroutine and start it running inside the multi-threaded scheduler */
```
The semantics are `setsched` which sets the scheduler for the coroutine. The coroutine is switched immediately to this new scheduler.  Any children coroutines to that coroutine automatically inherit the scheduler.

Some points about thread safety in no particular order:
- Thread safety is achieved using channels.
- Sockets in dsock are only thread-safe for generating sockets from listeners.

Some thoughts about coroutine introspection:
- A set of getters for internal information about current running coroutine, thread, stack size etc.

**Comment by sustrik** [comment]
I've been thinking about this some. Let me start with a cautionary tale:

In early 2000's SOA (CORBA/COM) was all the rage. The premise was that you can put interfaces between components in your program and then, when you need to scale it, you can magically turn those interfaces from local calls to RPCs. After almost a decade of hype the whole thing tumbled down because, of course, local calls and RPCs, although artificially similar, are in the end conceptually different beasts. There's several orders of magnitude performance difference, distributed systems have failure modes that monolithic systems don't (partial failure) and so on.

I think we are facing a similar problem here. It's tempting to assume that threads are conceptually equivalent to coroutines and abstract the difference away.  In other words, in the continuum of `machines --- processes --- threads --- coroutines` it's tempting to draw the line between processes and threads. In reality though the line rather lies between coroutines and threads. The reason being that beyond coroutines things actually start happening in parallel. (See also Robert Pike's famous 'concurrency is not parallelism' talk). 

The way I think about it is that coroutines are a simple language construct, similar to 'if', 'while' and such. Now, with 6ns context switch it's even true performance-wise. Threads, processes and machines all belong to a very different domain, the domain of distributed systems. The problems of distributed computing should, of course, be addressed, however, the problem is hard and nobody has good enough solutions yet.

What that means in practical terms is that in libdill we should provide the coroutine stuff. We should not try to do parallelism as there's no good one-fit-all solution yet and maybe there'll never be. Instead, we should define our interfaces in such a way that user can use distributed computing system of their choice and use libdill only for the 'coroutine as a language construct' stuff.

Now, I am guilty of falling into this trap myself. Introduction of proc() into libdill was just too tempting. Anyway, it was an error and I will remove it.

The good news is that we don't have to care about machines. They are nicely isolated and don't interfere with coroutines.

As for processes there's a little interference, e.g. sharing a listening socket among processes. We can get rid of that by having a separate listening process that would then pass connections to worker processes via AF_UNIX/SCM_RIGHTS.

As for threads, there's the inteference at the level of global variables. These are necessarily shared by all threads and lead to synchronization problems. Here's a list of global objects we have atm:

- main coroutine
- current coroutine
- list of ready coroutines
- list of pending timers
- handles array
- current pollset
- cached stacks
- census data

My idea is to pack all those into a single structure and provide and API to create/destroy such structures. (One will be created by default on process start-up so that people not caring about threads don't have to do any initialization.) The structure can then be attached to your thread (or whatever parallelization unit you have) e.g. by storing the pointer in TLS. There should also be an extension point to return appropriate context. This would then be used by all libdill functions instead of accessing the global state.

The above is, of course, bad news for work stealing. There's no way to take a coroutine and move it to different thread, same way as you can't move it to a different process or machine. One would have to do proper load-balancing to distribute the workload among cores.

**Comment by raedwulf** [comment]
> What that means in practical terms is that in libdill we should provide the coroutine stuff. We should not try to do parallelism as there's no good one-fit-all solution yet and maybe there'll never be. Instead, we should define our interfaces in such a way that user can use distributed computing system of their choice and use libdill only for the 'coroutine as a language construct' stuff.

Yep, in which case this is closer to becoming possible with my current WIP code.

> My idea is to pack all those into a single structure and provide and API to create/destroy such structures. (One will be created by default on process start-up so that people not caring about threads don't have to do any initialization.) The structure can then be attached to your thread (or whatever parallelization unit you have) e.g. by storing the pointer in TLS. There should also be an extension point to return appropriate context. This would then be used by all libdill functions instead of accessing the global state.

I was wondering about that - currently I've defined it so that the internal state is isolated per thread.
If we return a libdill `dill_context` (typical name for such a thing, unfortunate it clashes with the typical name for register contexts), it would need to be passed to each function and break the existing API.
I'm all for it and I guess it's better now than later.  I guess we can provide a compatibility API by creating a default structure which creates a default context that hides the extra parameter for the context.
In OpenGL, they do use the equivalent of Makecontext/setcurrentcontext to avoid adding this extra parameter although we need to consider the pros/cons against that model in libdill.  One benefit of Makecontext/setcurrentcontext is that the API is backwards compatible.

> The above is, of course, bad news for work stealing. There's no way to take a coroutine and move it to different thread, same way as you can't move it to a different process or machine. One would have to do proper load-balancing to distribute the workload among cores.

You are right here - also to add to the fact threads are not the same as cores.  I was thinking about work-stealing and load-balancing over threads when threads are again load-balanced by the operating system over different cores.  Though the normal assumption is if you want most performance threads = cores (different values is to try and adapt for inefficiencies of the OS scheduling/contention over resources).

> The above is, of course, bad news for work stealing. There's no way to take a coroutine and move it to different thread, same way as you can't move it to a different process or machine. One would have to do proper load-balancing to distribute the workload among cores.

In this model, it does become difficult to pry coroutines away from the current context.  I do have some thoughts about handling this but it is rather intrusive and I don't quite have a clear model in my head yet which will make it hard to explain.

**Comment by sustrik** [comment]
> it would need to be passed to each function and break the existing API

What about making it a compile-time option? E.g. set a option that disables the default context-getter, link with libdill and implement your own context-getter?

**Comment by sustrik** [comment]
Imagine we wanted to package the threading code with libdill proper. Without any compile-time options it would work as it does now. With --enable-threads there would be one new function dill_init(void) which would create the context and store the pointer in TLS.

Having that done we could reason about whether the threading code should be part of the core library and if not, how to decouple it.

**Comment by raedwulf** [comment]
Alright, we can do that :)

**Comment by sustrik** [comment]
Uh. And dill_term(void) to terminate the current context. We'll need to prevent leaks is user closes a thread.

**Comment by raedwulf** [comment]
Okay I've restructured my thoughts to consider the two problems that I hope libdill can solve:

1. I/O boundness of applications
2. CPU boundness of applications

In the first case, libdill is pretty good already; it does not block and transfers control to do something useful very quickly.

In the second case, libdill is good in a single thread and with thread-specific instances of libdill, can handle I/O-typical workloads which need to scale across multiple threads.

However, in the general case of CPU-bound work, there might be CPU-tasks that are short or long-running:
- Short running tasks are those that run quick enough that it doesn't make sense to ever swap out a short running task to distribute the load on the core.
- Long running tasks are those that run long enough that you might want to take it out and put it onto a core that is being starved of things to do.

In either case, we consider that rescheduling/migration across different threads only occurs when one thread is suspended and another thread is entering the scheduler.  This allows synchronisation to occur without interrupting the operation of a coroutine and have one of the threads move the necessary state over to the suspended thread. This rescheduling operation can actually occur in a coroutine that has a timer to wake up occasionally to check - like a watchdog.

This should hopefully prevents threads from being constantly blocked on I/O and being saturated with CPU-bounded work.

This leaves the problem of whether resources were being shared between coroutines prior to being migrated e.g. coroutine 1 depends on the same resources as coroutine 2, and coroutine 2 is being migrated.  My currently simplest solution would be documenting shareable resources, e.g. listener sockets and channels (if channels were made thread safe) and explicitly state other resources not being thread safe or use `hdup` to also track to ownership of resources to coroutines.

**Comment by raedwulf** [comment]
Despite not being in the typical use-case scenario of libdill, I think it's still an interesting one.
Games have different sets of work loads, one would be rendering, one would networking, one would be physics, and another for game logic.  Rendering is typically GPU-bound but can be file I/O bound as well to load resources from disk.  Networking is usually no problem on a client side, but game servers are notoriously badly optimised for networking. Physics is typically very CPU intensive (unless it runs on the GPU but that's more an exception than a rule). And game logic varies significantly from game to game, where it can be negligible to being more expensive in strategy games than anything else (though physics would probably not be present in strategy games).

Each of these are typically placed in different threads and this doesn't scale terribly well.  What has happened is that the workload is manually scheduled to a fixed number of threads.  What I believe is missing is the ability to split the work into different tasks and `yield` co-operatively between them so that at the yield-points the workload can be constantly balanced between the threads (and as a result across the different cores).  This is desired so that the work done by each thread completes before a soft-deadline of the next tick snapshot that is presented to the user.

Again the workloads vary signifcantly between different games; some games have all the data stored in memory and hopefully can fit it inside the GPU - so probably libdill is entirely unnecessary for those sorts of games.  However, larger open-world games require data to be constantly streamed from disk, and so there's is an I/O blocking element to that (I know we don't do asynchronous I/O over files, but on windows it is possible as well as using Linux's async API in a limited sense).  Furthermore, multiplayer games immediately have I/O blocking in terms of networking data to synchronise the game state of the server with the client.  Clients rarely have an issue, but game servers are notoriously bad at scaling with the number of clients.

**Comment by raedwulf** [comment]
I have a solution that doesn't require changing the current model of libdill and probably fits better into its philosophy.
What I hadn't got my head round was separating the concept of coroutine and actual work.  The idea is pre-create a number of coroutines at the start for each thread. At the start of a tick step, assign it a work task through a channel (all the coroutines block on a recv channel). The coroutines can follow different paths of logic because they can call different functions determined by a switch statement through the work task ID!

This allows the application/game to perform coarse-grain scheduling and manage the workload across different tick cycles! :)

**Comment by sustrik** [comment]
Your use case sounds completely reasonable. The problem is that I have no experience with that kind of thing. I guess people working on OS scheduling could be helpful with that kind of thing. So you are on your own here.

That being said, your last comment cuts directly into my area of expertise, that is, distributed systems. Let's have a look at it.

First, let's simplify it a bit. There's no need to preallocate coroutines, launching coroutines is cheap after all. What you need is a single "master" coroutine that runs in a loop, reads requests from a channel/socket, then does the big switch thing and launches an appropriate coroutine for each received request. If tasks are kept short, you can do what you call "coarse grained scheduling" and what is known in my world as "load balancing".

The challenge here is how to keep tasks evenly distributed among threads/processes/machines. That involves understanding the load at each thread/process/machine and pushing back (ceasing to receive more requests) if it exceeds some threshold.

The simplest way to estimate the load is to simply count the number of running tasks. That's basically what you propose in your comment.

More sophisticated way would be measure time spent doing stuff vs. time spent idling. You can then push back if load exceeds 90% or so.

**Comment by raedwulf** [comment]
I'll experiment once we get the thread-safety working but this sounds very promising.

Another throught - I'm looking at implementing a thread-safe IPC channel (which can be used cross-process as well).  In thinking about it, I was wondering what your current thoughts on having messages with priorities.  One use-case would be to ensure that control messages can be handled first over a single channel.  A second use-case would be to implement task priorities (in the above use-case). Priorities incur a little bit of overhead - most of which can be mitigated if only fixed number of levels are provided and a separate queue used for each one.  In fact, this potentially would allow high performance when multiple threads access it.

**Comment by sustrik** [comment]
Sorry for the delay. As for the prioritized queue, it' always implemented as a set of simple queues, one per priority level. So, in theory, you have to offer only a simple queue and the user can use multiple of them to handle priorities. This approach is being suggested to users by Golang team. That being said, the semantics of select/choose statement don't play particularly well with the multiple-queue approach. The code to receive messages from such a multi-queue is necessarily ugly. So, in the end, it can make sense to provide a prioritized queue out of the box.

**Comment by raedwulf** [comment]
Current tests pass: https://github.com/raedwulf/libdill/tree/threadsafe

I need to write a new test with pthread but it is looking close.

**Comment by sustrik** [comment]
That looks good. Once I have some I'll try pulling it into the mainline. I'll probably split the "global context" part and pull that in first. That'll make reviewing/reasoning about the thread part easier.

**Comment by raedwulf** [comment]
I'll close this soon if you're happy with the latest pull request.

**Comment by raedwulf** [comment]
There's a performance regression with `--enable-threads`... I'm currently investigating.  The regression only occurs with the shared library build (it's probably not able to inline something).

**Comment by sustrik** [comment]
The patch looks good. Let me know how's the investigation going.

**Comment by raedwulf** [comment]
I figured out what is causing it - the problem is that the compiler does not know how to optimise TLS accesses together without an extra level of indirection.  Adding that level of indirection slows down the single-threaded version.

I'm implementing some `#ifdefs` in a few places to make sure that single-threaded code has a fast path without that level of indirection i.e. it has statically allocated structures.

**Comment by sustrik** [comment]
I see. That's pretty anti-intuitive. Please, make sure to add a comment explaining the backgound to the code.

**Comment by raedwulf** [comment]
Have you seen cilk? It has slightly different paradigms to libdill, but it is interesting as it's also implemented into GCC.

**Comment by raedwulf** [comment]
This has been merged, closing this thread.
